### PR TITLE
Attribute #[warnings="warning-list"] to locally control warnings

### LINKED ISSUE #9430: [feature request] Ability to disable warnings locally
It would be great if there was a way to locally disable a warning -- similar to what `#[allow(foo)]` does in Rust, which disables the warning for the next item/statement only.

We currently have the "notation-overridden" warning disabled globally in Iris.  I'd like to enable it again, but we have some cases where overriding notation is just the most convenient thing to do:
```coq
Class Embed (A B : Type) := embed : A → B.
Arguments embed {_ _ _} _%I : simpl never.
Notation "⎡ P ⎤" := (embed P) : bi_scope.
Instance: Params (@embed) 3 := {}.
Typeclasses Opaque embed.

(* ... *)

Class BiEmbed (PROP1 PROP2 : bi) := {
  bi_embed_embed :> Embed PROP1 PROP2;
  bi_embed_mixin : BiEmbedMixin PROP1 PROP2 bi_embed_embed;
}.

(* ... *)

Section embed_laws.
  Context `{BiEmbed PROP1 PROP2}.
  (* Override notation to help type inference *)
  Local Notation embed := (embed (A:=PROP1) (B:=PROP2)).
  Local Notation "⎡ P ⎤" := (embed P) : bi_scope.
  Implicit Types P : PROP1.
```
This is actually a situation where overriding notation is (a) a local effect and (b) has predictable behavior, so I see no good reason to go through the pain of defining and opening a separate scope here.

I hear Coq has a notion of attributes now, seems like they would be a perfect fit for this.

### LINKED ISSUE #15893: Wish: attributes to disable warnings
Sometimes the desired way to deal with a warning is to locally disable it. This is particularly obvious for the several [warnings generated by importing MathComp](https://github.com/math-comp/math-comp/issues/90), but it is also true in other circumstances, such as overriding notations, declaring nonuniform coercions that don't have a uniform version, declaring a coercion that leads to ambiguous paths we don't care about, etc.

There is already a way to deal with this, which is to disable and enable the warnings surrounding the problematic lines:
```
Set Warnings "-notation-overridden, -ambiguous-paths".
From mathcomp Require Import all_ssreflect.
Set Warnings "notation-overridden, ambiguous-paths".
```

However, it would be nicer if there were attributes for disabling each warning (or a sensible subset of the warnings), such as the `nonuniform` one implemented recently in #15853.

### PR DESCRIPTION
**Pull request by SkySkimmer**
Close #15893
Close #9430

### CODE REVIEW

File: doc/sphinx/language/core/basic.rst [hunk 1]
```diff
@@ -437,6 +437,38 @@ boldface label "Attribute:".  Attributes are listed in the
       Set Warnings "unsupported-attributes".
       #[ foo ] Comments.
 
+Generic attributes
+++++++++++++++++++
```
Comment (Zimmi48) [inline doc/sphinx/language/core/basic.rst:441]: To fix the build:

```suggestion
^^^^^^^^^^^^^^^^^^
```

**Review by Zimmi48 — APPROVED** [review]

**Comment by SkySkimmer** [comment]
>Love that! To be nice to users, I wonder if we could support also the #[warning="..."] variant (no s). But only if it is trivial to do.

done

File: doc/changelog/08-vernac-commands-and-options/16902-warning-att.rst [hunk 2]
```diff
@@ -0,0 +1,5 @@
+- **Added:**
+  attributes :attr:`warnings` and alias :attr:`warning` to set warnings locally to a command
```
Comment (jfehrle) [inline doc/changelog/08-vernac-commands-and-options/16902-warning-att.rst:2]:
```suggestion
  attributes :attr:`warnings` and alias :attr:`warning` to set warnings locally for a command
```

"to" is not appropriate here.

File: doc/sphinx/language/core/basic.rst [hunk 3]
```diff
@@ -437,6 +437,43 @@ boldface label "Attribute:".  Attributes are listed in the
       Set Warnings "unsupported-attributes".
       #[ foo ] Comments.
 
+Generic attributes
+^^^^^^^^^^^^^^^^^^
+
+The following attribute is supported and handled in the same way by every command:
```
Comment (jfehrle) [inline doc/sphinx/language/core/basic.rst:443]:
```suggestion
The following attribute is supported  by every command:
```

File: doc/sphinx/language/core/basic.rst [hunk 4]
```diff
@@ -437,6 +437,43 @@ boldface label "Attribute:".  Attributes are listed in the
...
+
+   and `other-warning` is an error while executing the command.
+
+   Consequently, using this attribute around an :cmd:`Import` command
+   will prevent it from changing the warning state.
+
+   See also :opt:`Warnings`.
+
+.. attr:: warning = @string
+   :name: warning
```
Comment (jfehrle) [inline doc/sphinx/language/core/basic.rst:473]: :name: may not be needed.  I didn't check that, though.

File: doc/sphinx/language/core/basic.rst [hunk 5]
```diff
@@ -437,6 +437,43 @@ boldface label "Attribute:".  Attributes are listed in the
...
 
+Generic attributes
+^^^^^^^^^^^^^^^^^^
+
+The following attribute is supported and handled in the same way by every command:
+
+.. attr:: warnings = @string
+   :name: warnings
+
+   Sets the given warning string locally to the command. After the
```
Comment (jfehrle) [inline doc/sphinx/language/core/basic.rst:448]:
```suggestion
   Sets the given warning string locally for the command. After the
```

File: doc/sphinx/language/core/basic.rst [hunk 6]
```diff
@@ -437,6 +437,43 @@ boldface label "Attribute:".  Attributes are listed in the
...
+   command finishes the warning state is reset to what it was before
+   the command. For instance if the current warning state is
+   `some-warning,-other-warning`,
```
Comment (jfehrle) [inline doc/sphinx/language/core/basic.rst:451]:
```suggestion
   `some-warnings,-other-warning`,
```

To reflect that there is a set of warnings.

**Review by jfehrle — COMMENTED** [review]
Wording suggestions.  Remind me not to review things on my phone again :-).

File: doc/sphinx/language/core/basic.rst [hunk 7]
```diff
@@ -437,6 +437,43 @@ boldface label "Attribute:".  Attributes are listed in the
...
+
+      #[warnings="+other-warning"] Command.
+
+   is equivalent to
+
+   .. coqdoc::
+
+      Set Warnings "+other-warning".
+      Command.
+      Set Warnings "some-warnings,-other-warning".
```
Comment (jfehrle) [inline doc/sphinx/language/core/basic.rst:463]: Is some-warnings really needed?  IIUC unmentioned warnings are not changed and simpler for the reader to understand.

**Comment by SkySkimmer** [comment]
assignee?

File: doc/sphinx/language/core/basic.rst (hunk 7 above)
Comment (jfehrle) [reply doc/sphinx/language/core/basic.rst:463]: OK, but beyond the scope of this PR:  Does `Set Warnings "+my-warnings"` reset _all_ warnings?  The doc does not cover this detail.  If so, that seems unfriendly: if you have multiple cumulative `Set Warnings`, the second command has to repeat the settings from the first command (and similarly for the 3rd, ...)  IIUC this is different from how OCaml enables/disables warnings.

The doc for `Warnings` says + to make a warning an error and - to disable the warning.  Maybe that should be something like ! to make a warning an error, + to enable display of the warning (as a warning and not as an error) and - to disable displaying the warning.  And an entry without a prefix resets all warnings (if that's actually useful).

Comment (SkySkimmer) [reply doc/sphinx/language/core/basic.rst:463]: >beyond the scope of this PR

answered in zulip

**Comment by Alizter** [comment]
Also closes https://github.com/coq/coq/issues/9430

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
@coqbot run full ci

**Comment by Alizter** [comment]
@coqbot merge now

**Comment by Alizter** [comment]
@coqbot merge now

**Pull request merged**